src/Broker/Broker.php

```diff
@@ -228,11 +232,27 @@ public function hasConstant(\PhpParser\Node\Name $nameNode, Scope $scope): bool
 	 */
 	public function resolveConstantName(\PhpParser\Node\Name $nameNode, Scope $scope)
 	{
-		return $this->resolveName($nameNode, function (string $name): bool {
+		return $this->resolveName($nameNode, function (string $name) use ($scope): bool {
+			if ($name === '__COMPILER_HALT_OFFSET__' && $this->fileHasCompilerHaltStatementCalls($scope->getFile())) {
+				return true;
+			}
 			return defined($name);
 		}, $scope);
 	}
 
+	private function fileHasCompilerHaltStatementCalls(string $pathToFile): bool
+	{
+		//todo file is already parsed, it should not be parsed again
+		//todo I've just found no other way to check if __halt_compiler() statement was called in file
+		$parser = new DirectParser((new ParserFactory())->create(ParserFactory::PREFER_PHP7), new NodeTraverser());
```

Comment:
Just inject `Parser` instance into the Broker itself, don't instantiate it yourself. And it will inject cached implementation by default 😊